Update art3d.py to address strange behavior of depthshading on 3D scatterplots with close points

nhansendev opened the pull request:
## PR Summary
This update changes the depthshade behavior from the original normalization method to an improved version that behaves well even when points are closely spaced. This addresses issue [22861](https://github.com/matplotlib/matplotlib/issues/22861) 

I'm not clear on how to run the appropriate pytest unit tests for this file (if any), so I would appreciate some guidance there.

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

Obliman commented:
I think I've addressed the other issues, but am still getting image comparison failures. Is this to be expected when making changes related to how plots are displayed?

oscargus commented:
> Is this to be expected

Yes, it is expected. The RMS numbers are quite small, so only a number of pixels. 

I will let someone else have a proper say on this, but I believe that we should provide a fallback method for persons relying on the previous appearance. Unless it is obvious that the previous was always incorrect.

oscargus commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -488,6 +488,15 @@ def get_edgecolor(self):
         return self._maybe_depth_shade_and_sort_colors(super().get_edgecolor())
 
 
+def get_data_scale(X, Y, Z):
```
Please rename this to `_get_data_scale` to make it clear that it is a private method. A short doc-string would also be useful, possibly single-line.

oscargus commented:
If you please can provide the resulting figures from e.g. #22861 after applying the changes here, it would simplify the review.

oscargus commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -488,6 +488,15 @@ def get_edgecolor(self):
+            return 0
+        return np.power(max(data) - min(data), 2)
+
+    return np.power(_m(X) + _m(Y) + _m(Z), 0.5)
```
`np.sqrt`?

Obliman commented:
> If you please can provide the resulting figures from e.g. #22861 after applying the changes here, it would simplify the review.

I've attached some screenshots below showing the updated behavior.

This is with no alpha input, with depth-shading on. You can see how the alpha becomes uniform when depth is uniform (1st image), then fades proportionally to draw depth, including when draw depth reverses (2nd vs 3rd image):
![S2](https://user-images.githubusercontent.com/9289200/169644845-ff2f33ce-ad14-4940-84ec-7233724d8660.png)
![S1](https://user-images.githubusercontent.com/9289200/169644843-ad1a07f9-5803-4d7d-99be-e728ecb0548e.png)
![S3](https://user-images.githubusercontent.com/9289200/169644846-7b7d32da-087e-453b-926a-e48f1d8977d6.png)

When there is an alpha input (with depthshading on or off), the result is that the slight rotation of the plot no longer results in a sudden flip of alpha values as the depth order reverses:
![S4](https://user-images.githubusercontent.com/9289200/169645366-997ff6cd-367e-4ee8-b67d-9f0ee8949dba.png)
![S5](https://user-images.githubusercontent.com/9289200/169645368-a64cd5eb-f8e5-49b8-8a05-4c15229fe2a5.png)

Note that this behavior is only implemented for Path3DCollection since the bug is related to 3D scatterplots, but may be applicable to other collections if needed.

jklymak commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -918,15 +937,45 @@ def rotate_axes(xs, ys, zs, zdir):
+        # limits of the plot
+
+        # Solid near, transparent far, solid default
+        sats = np.clip(1 - (zs - min(zs)) / (_dscl+1e-8), 0.3, 1)
```
I'm not really following this, however, why 1e-8?  What if all your data contains small values order 1e-10?

Obliman replied:
I see your point. If the value of 1e-8 is too large relative to the data, then the values in sats will not cover the whole range of transparencies (0.3 to 1) and will instead be biased higher (e.g. 0.5 to 1). Do you have any suggestions for an alternative?

Obliman replied:
I'm realizing that I was only testing with well-behaved values and the calculation for sats is not as robust as I thought.

Obliman replied:
Aha! I remembered why I was doing it that way! It's a little misleading because "zs" sounds like a traditional z value, but it's actually the z-axis of the clipped _3D projection_ of the data, which ends up at a very consistent scale.

Example values for "zs" on a scatterplot of 10 points:
```
[-1.01084182 -1.00841226 -1.00599435 -1.003588   -1.00119315 -0.99880969
 -0.99643756 -0.99407667 -0.99172693 -0.98938829]
[-0.87861664 -0.90297352 -0.92871934 -0.95597639 -0.98488176 -1.01558962
 -1.048274   -1.08313208 -1.12038816 -1.1602985 ]
```

This _does_ vary, but I have never seen it come close to 1e-8 in testing. I forgot since I mainly worked on this a few weeks ago.

jklymak replied:
I guess if zs is normalized to be something visible, then 1e-8 is maybe OK?

Obliman replied:
The value is pretty arbitrarily set at the moment. Is there another value that you think would be more conservative?

jklymak replied:
Again, I'm not really familiar with where you are in the code - you are just trying to avoid a singularity, right?  Would `sats[~np.isfinite(sats)] = 0.3` work?

Obliman replied:
Thanks for the suggestions.

Unfortunately, while the method you're describing could handle the resulting NaN values, it would also result in warning messages because a zero-division would occur first:
```
/home/.../miniconda3/envs/matplotlib_debugging/testing.py:7: RuntimeWarning: divide by zero encountered in true_divide
  sats = np.clip(1 - (zs - min(zs)) / (dscl), 0.3, 1)
/home/.../miniconda3/envs/matplotlib_debugging/testing.py:7: RuntimeWarning: invalid value encountered in true_divide
  sats = np.clip(1 - (zs - min(zs)) / (dscl), 0.3, 1)
```

The goal is just to prevent zero-division when `_dscl==0`, which should only happen when the dataset is empty, or it contains exactly one point. Adding a small value in the denominator is one way to address this, but I suppose just checking the dataset length could work too.

jklymak replied:
...or checking `_dcs` is zero?

jklymak commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -488,6 +488,18 @@ def get_edgecolor(self):
+    Estimate the scale of the 3D data for use in depthshading
+    """
+    def _m(data):
+        if len(data) == 0:
```
Aren't x, y, z, are the same size at this point, so you don't need separate checks for len(data)==0?

Obliman replied:
Good point. Yes, I think they would all be the same size so only one check is needed.

Obliman commented:
I feel like I  should explain the comment "# Solid near, transparent far, solid default" further: depending on the equation used for calculating "sats" you can change the behavior of the depth-shading.

 The meaning is as follows:
Solid near: shallower datapoints are more/completely opaque
Transparent far: deeper datapoints become increasingly transparent 
Solid default: if everything is near equal depth then everything will be opaque

I felt that the other three versions were further from the desireable/existing depth-shading behavior, but could be interesting to include as alternative modes. 
Please let me know if you think this is a worthwhile inclusion.

QuLogic commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -918,15 +939,49 @@ def rotate_axes(xs, ys, zs, zdir):
 
 
-def _zalpha(colors, zs):
+def _zalpha(colors, zs, _dscl=None):
```
What is `_dscl`? No need to be stingy with letters in variable names.

Obliman replied:
How about "_data_scale", then?

QuLogic commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -488,6 +488,20 @@ def get_edgecolor(self):
+    Estimate the scale of the 3D data for use in depthshading
+    """
+    def _m(data):
+        return np.power(max(data) - min(data), 2)
```
```suggestion
        return np.ptp(data)**2
```
and maybe this is short enough to inline.

Obliman replied:
Nice, I haven't seen that function before. I'll make it inline.

QuLogic commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -488,6 +488,20 @@ def get_edgecolor(self):
 
+def _get_data_scale(X, Y, Z):
+    """
+    Estimate the scale of the 3D data for use in depthshading
```
```suggestion
    Estimate the scale of the 3D data for use in depth shading.
```

QuLogic commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -612,14 +627,20 @@ def do_3d_projection(self):
             color_array = color_array[self._z_markers_idx]
-        return mcolors.to_rgba_array(color_array, self._alpha)
+
+        # The correct alpha is already accounted for in the color_array
+        # Passing the self._alpha value in overwrites the automatically
+        # adjusted alpha
```
This comment is about code that previously existed, which is confusing; please reword so it's about the current code.

Obliman replied:
Will remove since no comment was previously needed. I was just noting why it was changed.

QuLogic commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -918,15 +939,49 @@ def rotate_axes(xs, ys, zs, zdir):
+        # Normalize the z-depths to the range 0 - 1
+        norm = Normalize(min(zs), max(zs))
+
+        # Generate alpha multipliers using the normalized z-depths so that
+        # closer points
+        # are opaque and the furthest points are still visible, but transparent
```
Re-wrap.

Obliman replied:
Will do

QuLogic commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -918,15 +939,49 @@ def rotate_axes(xs, ys, zs, zdir):
+        # Improved normalization using a scale value derived from the XYZ
+        # limits of the plot
+
+        # Solid near, transparent far, solid default
+        if _dscl == 0:
+            sats = np.ones_like(zs)
+        else:
+            sats = np.clip(1 - (zs - min(zs)) / _dscl, 0.3, 1)
+        # print(_dscl)
+
+        # Solid near, transparent far, transparent default
+        # sats = np.clip((max(zs)-zs)/(_dscl+1e-8), 0.3, 1)
+
+        # Transparent near, solid far, solid default
+        # sats = np.clip(1-(max(zs)-zs)/(_dscl+1e-8), 0.3, 1)
+
+        # Transparent near, solid far, transparent default
+        # sats = np.clip((zs-min(zs))/(_dscl+1e-8), 0.3, 1)
```
Yes, not sure what `Solid near, transparent far, solid default` means, especially with these different options.

Obliman replied:
I don't think we've reached a conclusion on whether it's worthwhile implementing all the different modes, or just the one that approximates the previous behavior. Either way, here's a cleaned-up version of the function that improves on the comments:

```

def _zalpha(colors, zs, _data_scale=None, shading_mode=0):
    """Modify the alphas of the color list according to depth."""
    # Return empty if given empty
    if len(colors) == 0 or len(zs) == 0:
        return np.zeros((0, 4))

    if _data_scale is None:
        # This only works well if the points for *zs* are well-spaced
        # in all three dimensions. Otherwise, at certain orientations,
        # the min and max zs are very close together.
        # Should really normalize against the viewing depth.

        # Normalize the z-depths to the range 0 - 1
        norm = Normalize(min(zs), max(zs))

        # Generate alpha multipliers using the normalized z-depths so that
        # closer points are opaque and the furthest points are still visible,
        # but transparent
        sats = 1 - norm(zs) * 0.7

    else:
        # Improved normalization using a scale value derived from the XYZ
        # limits of the plot

        if _data_scale == 0:
            # Don't scale the alpha values since we have no valid
            # data scale for reference
            sats = np.ones_like(zs)

        else:
            if shading_mode == 0:
                # This is the mode that most closely matches the behavior
                # of when _data_scale=None, and legacy code

                # Shallower points have an increasingly solid appearance
                # Deeper points have an increasingly transparent appearance
                # Points with uniform depth have a solid appearance
                sats = np.clip(1 - (zs - min(zs)) / _data_scale, 0.3, 1)

            elif shading_mode == 1:
                # Shallower points have an increasingly solid appearance
                # Deeper points have an increasingly transparent appearance
                # Points with uniform depth have a transparent appearance
                sats = np.clip((max(zs) - zs) / _data_scale, 0.3, 1)

            elif shading_mode == 2:
                # Shallower points have an increasingly transparent appearance
                # Deeper points have an increasingly solid appearance
                # Points with uniform depth have a solid appearance
                sats = np.clip(1 - (max(zs) - zs) / _data_scale, 0.3, 1)

            elif shading_mode == 3:
                # Shallower points have an increasingly transparent appearance
                # Deeper points have an increasingly solid appearance
                # Points with uniform depth have a transparent appearance
                sats = np.clip((zs - min(zs)) / _data_scale, 0.3, 1)

            else:
                raise NotImplementedError('Specified "shading_mode" must be 0 (default), 1, 2, or 3')

    rgba = np.broadcast_to(mcolors.to_rgba_array(colors), (len(zs), 4))

    # Change the alpha values of the colors using the generated alpha
    # multipliers
    return np.column_stack([rgba[:, :3], rgba[:, 3] * sats])
```

I would just have to add the kwarg to upstream function calls to give the user access.
Please let me know if there's a better/preferred way to handle when the shading mode is incorrectly specified (the NotImplementedError).

QuLogic commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -918,15 +939,49 @@ def rotate_axes(xs, ys, zs, zdir):
+        # Transparent near, solid far, transparent default
+        # sats = np.clip((zs-min(zs))/(_dscl+1e-8), 0.3, 1)
+
+    # Restructure colors into a rgba numpy array
```
This comment is somewhat redundant.

Obliman replied:
Agreed, will remove

QuLogic commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -918,15 +939,49 @@ def rotate_axes(xs, ys, zs, zdir):
-    #        in all three dimensions. Otherwise, at certain orientations,
-    #        the min and max zs are very close together.
-    #        Should really normalize against the viewing depth.
+    # Return empty if given empty
```
Redundant comment.

Obliman replied:
Agreed, will remove

QuLogic commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -918,15 +939,49 @@ def rotate_axes(xs, ys, zs, zdir):
+    # Restructure colors into a rgba numpy array
     rgba = np.broadcast_to(mcolors.to_rgba_array(colors), (len(zs), 4))
+
+    # Change the alpha values of the colors using the generated alpha
+    # multipliers
```
Also somewhat redundant.

Obliman replied:
I would prefer to leave it in just to be explicit. With the indexing and column stacking the intent might not be obvious at a glance

QuLogic reviewed (commented):
I find some of the comments redundant (as noted); they should describe _why_, not _what_.

airalcorn2 commented:
Thanks for this. This behavior was driving me crazy.

tacaswell commented:
This is cool work!  A couple of requests:

 - can we use an enum (or strings) rather than integers for the modes?
 - can you add a whats_new to show this off?
 - can you add tests?

I have pushed this to the 3.8 milestone.

Obliman commented:
> * can we use an enum (or strings) rather than integers for the modes?

Definitely, though I'm not sure what appropriate string representations would be. My previous convention was to summarize behavior as follows: `"STS": Solid when shallow, Transparent when deep, Solid default`, `"STT": Solid when shallow, Transparent when deep, Transparent default`, etc.

> * can you add a whats_new to show this off?
> * can you add tests?

Are there guidelines for these/examples I could reference?

Thanks for the suggestions.

melissawm commented:
Hi @Obliman - For the tests, you can take a look at [this file](https://github.com/matplotlib/matplotlib/blob/main/lib/mpl_toolkits/mplot3d/tests/test_art3d.py) which contains the tests for `art3d.py`. For the what's new, here's [the documentation](https://matplotlib.org/stable/users/next_whats_new/README.html). Let me know if you have other questions. Cheers!

jklymak commented:
Noted as draft until tests are fixed.  Thanks!

scottshambaugh commented:
Hi @nhansendev, are you still interested in finishing up this bugfix/feature? Thanks!

nhansendev commented:
> Hi @nhansendev, are you still interested in finishing up this bugfix/feature? Thanks!

Thanks for the reminder, I've made some updates.

I've thought about the changes I had proposed and the whole "different modes" thing was just unnecessarily complicated/probably not that useful. I've since simplified it to be three kwargs:

1. inverted:     whether to progress from opaque to transparent (default) or the opposite
2. min_alpha: the smallest allowable alpha value (highest transparency)
3. legacy:       whether to use the old algorithm

I've added the "what's new" document and updated other classes to provide access to the user when creating scatter plots or otherwise interacting with Patch3DCollection/Path3DCollection objects.

oscargus commented:
Yes, I think that there should be some tests for this to make sure that possible later changes does not break the behavior.

One suggestion is to add an image comparison test with a number of subplots, each subplot illustrating some combination of the new behavior. 

Looking at the code coverage, I can imagine a 2 x 2 subplot where each combination of legacy/new and inverted/non-inverted is plotted. (Right now, only new and non-inverted seems to be tested.)

For min-alpha one can consider something similar or simply checking the alphas in the resulting collection to confirm that they are as expected (if one can avoid image tests that is slightly better since storing images leads to a rather large repo...).

oscargus commented:
Oh, and for "completeness", one can also call the new `get_*` methods as part of some test. Simply confirming that they return the expected value.

tacaswell commented:
This is breaking tests that were expecting the old behavior.  We either need to manually set the legacy behavior in the tests and add an API change note or we need to set the default to set the default value of `depthshade_legacy` to `True`.

Rather than adding a bunch of parameters named `depthshade_XXX`, we should think of ways to collapse them (either bool + dict or relying on the type of `depthshade` (suggeted by @ksunden ).

Moving this to 3.10 and should talk about this on the call as soon as 3.9 is branched.

scottshambaugh commented:
Hi @nhansendev, are you still interested in finishing up this PR?

nhansendev commented:
> Hi @nhansendev, are you still interested in finishing up this PR?

Hey, sorry for the delay in response. If someone else wants to tackle the last steps I would appreciate it.

scottshambaugh commented:
Sounds good @nhansendev, thank you for the work getting it to this point! I'll finish up the remaining items in https://github.com/matplotlib/matplotlib/pull/29287

scottshambaugh commented:
Commenting on this PR to notify original authors - what is the use case behind the `inverted` method? I don't immediately see where it would be useful.

My personal leaning here is to consider the current / `legacy` behavior a bug that is appropriate to change without backwards compatibility rather than part of the ["visual API"](https://matplotlib.org/devdocs/devel/api_changes.html#add-or-change-colormaps-color-sequences-and-styles) we should not break. Removing both those options would simplify the implementation.

nhansendev commented:
> Commenting on this PR to notify original authors - what is the use case behind the `inverted` method? I don't immediately see where it would be desired.

This was trivial to implement, so was added in case someone finds it useful. If it's just adding more work then I think it can be removed.

> My personal leaning here is to consider the current / `legacy` behavior a bug that is appropriate to change without backwards compatibility 

I agree that the legacy behavior represents an undesirable visual bug. Keeping backwards compatibility was suggested earlier, but it's _probably_ not going to significantly affect generated plots given it was barely failing the old image tests.

scottshambaugh commented:
Fix is merged in, thank you again for the initial work on this!

nhansendev commented:
> Fix is merged in, thank you again for the initial work on this!

Thank you for finishing it 🥇
